### PR TITLE
async support, timestamps, interface unification

### PR DESCRIPTION
**Pull request by bartes**
resolves #22 

syncing interface with ruby client (https://github.com/castle/castle-ruby/)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,10 @@
 *.pyc
 *.pyo
 *.egg-info
+.eggs
+.pytest_cache
 /dist
 /build
+/.eggs
```
Comment (nijikon) [inline .gitignore]: Duplicated

File: castle/client.py [hunk 2]
```diff
@@ -7,39 +7,71 @@
...
+        if self.timestamp:
+            options.setdefault(self.timestamp)
+
     def authenticate(self, options):
         if self.tracked():
+            self._add_timestamp_if_necessary(options)
+            command = CommandsAuthenticate(self.context).build(options)
             try:
-                response = self.api.call(CommandsAuthenticate(self.context).build(options))
+                response = self.api.call(command)
```
Comment (nijikon) [inline castle/client.py]: No need to create `command` variable, just pass it directly.

File: castle/context/sanitizer.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class ContextSanitizer(object):
+
```
Comment (nijikon) [inline castle/context/sanitizer.py]: Empty line

File: castle/test/utils_test.py [hunk 4]
```diff
@@ -99,3 +102,12 @@ def test_deep_merge(self):
             }
         }
         self.assertEqual(a, expected)
+
+
+class TimestampTestCase(unittest.TestCase):
+
```
Comment (nijikon) [inline castle/test/utils_test.py]: Empty line

File: castle/test/validators/not_supported_test.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+from castle.test import unittest
+from castle.validators.not_supported import ValidatorsNotSupported
+from castle.exceptions import InvalidParametersError
+
+class ValidatorsNotSupportedTestCase(unittest.TestCase):
+
```
Comment (nijikon) [inline castle/test/validators/not_supported_test.py]: Empty line

File: castle/test/validators/present_test.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+from castle.test import unittest
+from castle.validators.present import ValidatorsPresent
+from castle.exceptions import InvalidParametersError
+
+class ValidatorsPresentTestCase(unittest.TestCase):
+
```
Comment (nijikon) [inline castle/test/validators/present_test.py]: Empty line

File: castle/validators/not_supported.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+from castle.exceptions import InvalidParametersError
+
+
+class ValidatorsNotSupported(object):
+
```
Comment (nijikon) [inline castle/validators/not_supported.py]: Empty line

File: castle/validators/present.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+from castle.exceptions import InvalidParametersError
+
+
+class ValidatorsPresent(object):
+
```
Comment (nijikon) [inline castle/validators/present.py]: Empty line

File: castle/client.py (hunk 2 above)
Comment (bartes) [reply castle/client.py]: @nijikon - I wanted to remove that from try block

File: castle/utils.py [hunk 9]
```diff
@@ -1,20 +1,31 @@
...
+
     :param base: The base dictionary which will be merged into.
     :param extra: The dictionary to merge into the base. Keys from this
         dictionary will take precedence.
     """
-    for key in extra:
+    if extra is None:
+        return
+
+    for key, value in extra.items():
```
Comment (jprestel-rue) [inline castle/utils.py]: because you're supporting python 2 and 3, I would add a dependency on `six` and do

```python
from six import iteritems
...
for key, value in iteritems(extra)
```

File: castle/validators/present.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+    @staticmethod
+    def call(options, *args):
+        for key in args:
+            if options.get(key) is None or options.get(key) == '':
+                raise InvalidParametersError("{key} is missing or empty".format(key=key))
```
Comment (jprestel-rue) [inline castle/validators/present.py]: do you need to test for empty array or empty object values?

File: setup.py [hunk 11]
```diff
@@ -27,7 +27,9 @@
         'Programming Language :: Python :: 3.5',
         'Programming Language :: Python :: 3.6',
     ],
-    install_requires=['requests>=2.5'],
+    install_requires=[
+        'requests>=2.5',
+    ],
     tests_require=['responses'],
```
Comment (jprestel-rue) [inline setup.py]: did you consider my alternate method of specifying test requirements? it makes it so you can easily install test dependencies by running `pip install -e .[test]`

https://github.com/castle/castle-python/pull/22/commits/f044a6bd71cd9c30175f4153413ff45f0cccec4b

File: castle/utils.py [hunk 12]
```diff
@@ -1,20 +1,31 @@
...
+            deep_merge(base[key], value)
+        else:
+            # Otherwise, set the key on the base to be the value of the extra.
+            base[key] = value
 
-        # Otherwise, set the key on the base to be the value of the extra.
-        base[key] = extra[key]
+def timestamp():
+    """Return an ISO8601 timestamp representing the current datetime in UTC."""
+    return datetime.utcnow().isoformat()[:-3]
```
Comment (jprestel-rue) [inline castle/utils.py]: why truncate the microseconds?

also, on the off chance that the timestamp is generated where microseconds=0, `isoformat` actually omits them entirely. In that case, your slice will actually lop off the seconds.

```python
>>> from datetime import datetime
>>> d = datetime.utcnow().replace(microsecond=0)
>>> d.isoformat()
'2018-02-08T17:34:53'
>>> d.isoformat()[:-3]
'2018-02-08T17:34'
```

Comment (jprestel-rue) [reply castle/utils.py]: Separately, the timestamp lacks the timezone offset. While we know we requested the timestamp in UTC, sending it over the wire this way is ambiguous and feels like the wrong thing to do. The simplest thing to do would be to tack on  a `+00:00` to the end of the string.

File: castle/utils.py [hunk 13]
```diff
@@ -1,20 +1,31 @@
...
+    """
+    Deeply merge two dictionaries, overriding existing keys in the base.
+
     :param base: The base dictionary which will be merged into.
     :param extra: The dictionary to merge into the base. Keys from this
         dictionary will take precedence.
     """
-    for key in extra:
+    if extra is None:
+        return
```
Comment (jprestel-rue) [inline castle/utils.py]: should add a test for this

File: castle/utils.py [hunk 14]
```diff
@@ -1,20 +1,31 @@
...
+
+    for key, value in extra.items():
         # If the key represents a dict on both given dicts, merge the sub-dicts
-        if key in base and isinstance(base[key], dict)\
-                and isinstance(extra[key], dict):
-            deep_merge(base[key], extra[key])
-            continue
+        if value is None:
+            if key in base:
+                del base[key]
```
Comment (jprestel-rue) [inline castle/utils.py]: should add a test for this behavior

File: castle/validators/present.py [hunk 15]
```diff
@@ -0,0 +1,10 @@
+from castle.exceptions import InvalidParametersError
+
+
+class ValidatorsPresent(object):
+
+    @staticmethod
+    def call(options, *args):
```
Comment (jprestel-rue) [inline castle/validators/present.py]: FWIW, I find this pattern of static `call` method on a class confusing. It seems to me that all the class is doing is namespacing the function... which I don't see any benefit to. It feels like this would be simpler as:
```python
def validate_present(options, *keys):
```

File: castle/extractors/ip.py [hunk 16]
```diff
@@ -9,6 +9,11 @@ def call(self):
         return self.wsgi_request_ip()
 
     def wsgi_request_ip(self):
+        cf_ip = self.request.environ.get('HTTP_CF_CONNECTING_IP')
```
Comment (jprestel-rue) [inline castle/extractors/ip.py]: I had to lookup what this header is.  I would add a comment indicating what it is

File: castle/context/default.py [hunk 17]
```diff
@@ -31,6 +31,14 @@ def defaults_extra(self):
...
-        context = dict(self.defaults())
-        context.update(self.defaults_extra())
+        context = dict(self._defaults())
+        context.update(self._defaults_extra())
         return context
+
+    @staticmethod
+    def _fetch_cookies(request, cookies):
+        if cookies:
+            return cookies
```
Comment (jprestel-rue) [inline castle/context/default.py]: the previous behavior was to use the passed in cookies only if the request cookies didn't exist. this doesn't feel like a bad change, just wanted to call it out.

File: castle/configuration.py [hunk 18]
```diff
@@ -12,7 +12,9 @@
     'Forwarded',
     'X-Forwarded',
     'X-Real-IP',
-    'REMOTE_ADDR'
+    'REMOTE_ADDR',
+    'X-Forwarded-For',
```
Comment (jprestel-rue) [inline castle/configuration.py]: `X-Forwarded-For` is in the list twice

File: castle/test/extractors/ip_test.py [hunk 19]
```diff
@@ -5,21 +5,31 @@
...
     req.environ = {'REMOTE_ADDR': request_ip()}
     return req
 
+
 def request_without_ip_x_forwarded_for():
     req = mock.Mock(spec=['environ'])
     req.environ = {'HTTP_X_FORWARDED_FOR': request_ip()}
     return req
 
+def request_without_ip_cf_connecting_ip():
```
Comment (jprestel-rue) [inline castle/test/extractors/ip_test.py]: i think you mean `with` not `without` for this and the functions above it.

File: castle/test/client_test.py [hunk 20]
```diff
@@ -103,58 +110,46 @@ def test_track_tracked_true(self):
...
-        client = Client(request(), {})
+        client = Client.from_request(request(), {})
         client.disable_tracking()
         self.assertEqual(client.do_not_track, True)
         client.enable_tracking()
         self.assertEqual(client.do_not_track, False)
 
-    def test_tracked_true(self):
-        client = Client(request(), {})
+    def test_dont_tracked_false(self):
```
Comment (jprestel-rue) [inline castle/test/client_test.py]: i think you want `test_do_not_track_false`

File: castle/test/client_test.py [hunk 21]
```diff
@@ -103,58 +110,46 @@ def test_track_tracked_true(self):
...
+        client = Client.from_request(request(), {})
         self.assertEqual(client.tracked(), True)
 
-    def test_tracked_false(self):
-        client = Client(request(), {})
-        client.disable_tracking()
+    def test_dont_tracking_true(self):
```
Comment (jprestel-rue) [inline castle/test/client_test.py]: i think you want `test_do_not_track_true`

File: castle/test/commands/authenticate_test.py [hunk 22]
```diff
@@ -1,89 +1,85 @@
...
+        # patch timestamp to return a known value
+        timestamp_patcher = mock.patch('castle.commands.authenticate.timestamp')
+        self.mock_timestamp = timestamp_patcher.start()
+        self.mock_timestamp.return_value = mock.sentinel.timestamp
+        self.addCleanup(timestamp_patcher.stop)
+
     def test_init(self):
-        self.assertIsInstance(CommandsAuthenticate({}).context_merger, ContextMerger)
+        obj = CommandsAuthenticate({})
+        self.assertEqual(obj.context, {})
```
Comment (jprestel-rue) [inline castle/test/commands/authenticate_test.py]: I think this would be a more effective test if you used a sentinel:
```python
context = mock.sentinel.test_init_context
obj = CommandsAuthenticate(context)
self.assertEqual(obj.context, context)
```

File: castle/utils.py (hunk 9 above)
Comment (bartes) [reply castle/utils.py]: with sdks I prefer to limit dependencies to total minimum (zero preferred). we can discuss performance improvements in next PR

File: castle/validators/present.py (hunk 10 above)
Comment (bartes) [reply castle/validators/present.py]: we need to validate options for non empty values specified by args

File: setup.py (hunk 11 above)
Comment (bartes) [reply setup.py]: yes I reverted that because I had troubles to make it working with travis (we can try to add it in the next PR with proper changes to CI config)

File: castle/utils.py (hunk 12 above)
Comment (bartes) [reply castle/utils.py]: we are not removing microseconds `replace(microsecond=0)` in that method so they will be properly truncated and delivered to the api. But I totally agree with you. we should send full iso format no matter how it looks like, to make it more clear and readable and api will take care of proper formatting

File: castle/validators/present.py (hunk 15 above)
Comment (bartes) [reply castle/validators/present.py]: we plan to have similar interface in all the sdks

File: castle/extractors/ip.py (hunk 16 above)
Comment (bartes) [reply castle/extractors/ip.py]: Valid point. I plan to extend README with all the info

File: castle/context/default.py (hunk 17 above)
Comment (bartes) [reply castle/context/default.py]: the behavior is opposite. to prevent reading from the request you can pass similar information as an argument (just in case there will be some malformed or different kind of request object there)

File: castle/test/client_test.py (hunk 20 above)
Comment (bartes) [reply castle/test/client_test.py]: ahh that naming .. will review that and correct 👍

**Comment by bartes** [comment]
thanks @jprestel-rue . great feedback

File: castle/validators/not_supported.py (hunk 7 above)
Comment (bartes) [reply castle/validators/not_supported.py]: @nijikon - I used autopep8 for empty lines styling

File: castle/validators/present.py (hunk 10 above)
Comment (jprestel-rue) [reply castle/validators/present.py]: what i meant is that empty array `[]` and empty dict `{}` are typically considered empty values like a blank string is `''`

Comment (bartes) [reply castle/validators/present.py]: got it.. yes this is a very basic validation, just to handle the most obvious cases.. the api will return an error if wrong format will be provided like an array instead of a string.. but this is a great idea for a next enhancement  (I added that idea to sdk backlog )
